internal/controlplane/handlers_repositories.go

```diff
@@ -308,29 +309,61 @@ func (s *Server) ListRemoteRepositoriesFromProvider(
 	entityCtx := engine.EntityFromContext(ctx)
 	projectID := entityCtx.Project.ID
 
-	provider, err := getProviderFromRequestOrDefault(ctx, s.store, in, projectID)
+	provs, err := listProvidersOrInferDefault(ctx, s.store, in, projectID)
 	if err != nil {
 		return nil, providerError(err)
 	}
 
-	zerolog.Ctx(ctx).Debug().
-		Str("provider", provider.Name).
-		Str("projectID", projectID.String()).
-		Msg("listing repositories")
-
-	pbOpts := []providers.ProviderBuilderOption{
-		providers.WithProviderMetrics(s.provMt),
-		providers.WithRestClientCache(s.restClientCache),
-	}
-	p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+	out := &pb.ListRemoteRepositoriesFromProviderResponse{
+		Results: []*pb.UpstreamRepositoryRef{},
 	}
 
-	if !p.Implements(db.ProviderTypeRepoLister) {
-		return nil, util.UserVisibleError(codes.Unimplemented, "provider does not implement repository listing")
+	failIfListerNotImplemented := len(provs) == 1
+
+	for _, provider := range provs {
+		zerolog.Ctx(ctx).Debug().
+			Str("provider", provider.Name).
+			Str("projectID", projectID.String()).
+			Msg("listing repositories")
+
+		pbOpts := []providers.ProviderBuilderOption{
+			providers.WithProviderMetrics(s.provMt),
+			providers.WithRestClientCache(s.restClientCache),
+		}
+		p, err := providers.GetProviderBuilder(ctx, provider, s.store, s.cryptoEngine, &s.cfg.Provider, pbOpts...)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "cannot get provider builder: %v", err)
+		}
+
+		// check if the provider implements the repo lister interface
+		if failIfListerNotImplemented && !p.Implements(db.ProviderTypeRepoLister) {
+			return nil, util.UserVisibleError(codes.Unimplemented, "provider does not implement repository listing")
+		} else if !p.Implements(db.ProviderTypeRepoLister) {
+			// skip the provider if it does not implement the repo lister interface
+			continue
+		}
+
+		results, err := s.listRemoteRepositoriesForProvider(ctx, provider.Name, p, projectID)
+		if err != nil {
+			return nil, err
+		}
+
+		out.Results = append(out.Results, results...)
+
+		// Telemetry logging
+		logger.BusinessRecord(ctx).Project = projectID
```

Comment:
I don't think we want to repeatedly set the projectID to be the same thing on each call.

Additionally, I think we want to set `.Project` _before_ this loop (as early as we know it), so that we get RPC telemetry even for RPC failures.  (I missed this during the initial telemetry pass, and I've been commenting as we refactor the RPC methods.)